tests/core-addon/IonCore.test.js

```diff
@@ -19,7 +20,138 @@ describe('IonCore', function () {
+    it('rejects unknown messages', function () {
+      // Mock the URL of the options page.
+      const TEST_OPTIONS_URL = "install.sample.html";
+      chrome.runtime.getURL.returns(TEST_OPTIONS_URL);
+
+      // Provide an unknown message type and a valid origin:
+      // it should fail due to the unexpected type.
+      assert.rejects(
+        this.ionCore._handleMessage(
+          {type: "test-unknown-type", data: {}},
+          {url: TEST_OPTIONS_URL}
+        ),
+        { message: "IonCore - unexpected message type test-unknown-type"}
+      );
+    });
+
+    it('rejects unknown senders', function () {
+      // Mock the URL of the options page.
+      const TEST_OPTIONS_URL = "install.sample.html";
+      chrome.runtime.getURL.returns(TEST_OPTIONS_URL);
+
+      // Provide an unknown message type and a valid origin:
+      // it should fail due to the unexpected type.
+      assert.rejects(
+        this.ionCore._handleMessage(
+          {type: "enroll", data: {}},
+          {url: "unkown-sender-url.html"}
+        ),
+        { message: "IonCore - received message from unexpected sender"}
+      );
+    });
+
+    it('dispatchers enrollment messages', async function () {
+      // Mock the URL of the options page.
+      const TEST_OPTIONS_URL = "install.sample.html";
+      chrome.runtime.getURL.returns(TEST_OPTIONS_URL);
+
+      // Create a mock for the telemetry API.
+      const FAKE_UUID = "c0ffeec0-ffee-c0ff-eec0-ffeec0ffeec0";
+      chrome.legacyTelemetryApi = {
+        generateUUID: async function() { return FAKE_UUID; },
+        setIonID: async function(uuid) {},
+        submitEncryptedPing: async function(type, payload, options) {},
+      };
+      let telemetryMock = sinon.mock(chrome.legacyTelemetryApi);
+      // Use the spy to record the arguments of submitEncryptedPing.
+      let telemetrySpy =
+        sinon.spy(chrome.legacyTelemetryApi, "submitEncryptedPing");
+      // Make sure to mock the local storage calls as well.
+      chrome.storage.local.set.yields();
+
+      // Provide a valid enrollment message.
+      await this.ionCore._handleMessage(
+        {type: "enrollment", data: {}},
+        {url: TEST_OPTIONS_URL}
+      );
+
+      // We expect to store the fake ion ID...
+      telemetryMock.expects("setIonID").withArgs([FAKE_UUID]).calledOnce;
+      // ... to submit a ping with the expected type ...
+      const submitArgs = telemetrySpy.getCall(0).args;
+      assert.equal(submitArgs[0], "pioneer-study");
+      // ... an empty payload ...
+      assert.equal(Object.keys(submitArgs[1]).length, 0);
+      // ... and a specific set of options.
+      assert.equal(submitArgs[2].studyName, "pioneer-meta");
+      assert.equal(submitArgs[2].encryptionKeyId, "discarded");
+      assert.equal(submitArgs[2].schemaName, "pioneer-enrollment");
+      assert.equal(submitArgs[2].schemaNamespace, "pioneer-meta");
+    });
+
+    it('dispatchers study-enrollment messages', async function () {
```

Comment:
```suggestion
    it('dispatches study-enrollment messages', async function () {
```

Reply:
GAH! Good catch :-) I'll address this right after reviews are complete, so that I can fix everything in one go!

Reply:
You can commit directly in this UI! Didn't want to make you do more work :)